Heredoc delimiter/marker not parsed correctly

Heredocs made by perverse people where the delimiter contains the `'` or `"` characters are not parsed correctly. I'm not sure if it's valid code, but it works in Busybox's ash and Bash. The [posix docs](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_04) say
> If any part of word is quoted, the delimiter shall be formed by performing quote removal on word, and the here-document lines shall not be expanded.


#### For bugs
- Rule Ids: SC1073, SC1072, SC1042, SC1041
- version: 0.6.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC1041)
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh

cat <<'"EOF'
hello
"EOF
```

likewise for 
```sh
#!/bin/sh

cat <<'E'\''OF'
hello
E'OF
```

#### Here's what shellcheck currently says:

> Line 3:
> cat <<'"EOF'
> ^-- SC1009: The mentioned syntax error was in this simple command.
>     ^-- SC1073: Couldn't parse this here document. Fix to allow more checks.
>       ^-- SC1041: Found 'EOF' further down, but not on a separate line.
>       ^-- SC1042: Close matches include '"EOF' (!= 'EOF').
>  
> Line 6:
> 
> ^-- SC1072: Here document was not correctly terminated. Fix any mentioned problems and try again.

#### Here's what I wanted or expected to see:

> No issues detected!

--- a/src/ShellCheck/Parser.hs
+++ b/src/ShellCheck/Parser.hs
@@ -1739,6 +1739,7 @@ prop_readHereDoc14= isWarning readScript "cat << foo\nbar\nfoo \n"
 prop_readHereDoc15= isWarning readScript "cat <<foo\nbar\nfoo bar\nfoo"
 prop_readHereDoc16= isOk readScript "cat <<- ' foo'\nbar\n foo\n"
 prop_readHereDoc17= isWarning readScript "cat <<- ' foo'\nbar\n  foo\n foo\n"
+prop_readHereDoc18= isOk readScript "cat <<'\"foo'\nbar\n\"foo\n"
 prop_readHereDoc20= isWarning readScript "cat << foo\n  foo\n()\nfoo\n"
 prop_readHereDoc21= isOk readScript "# shellcheck disable=SC1039\ncat << foo\n  foo\n()\nfoo\n"
 readHereDoc = called "here document" $ do
@@ -1759,14 +1760,17 @@ readHereDoc = called "here document" $ do
     addPendingHereDoc doc
     return doc
   where
-    quotes = "\"'\\"
+    unquote :: String -> (Quoted, String)
+    unquote "" = (Unquoted, "")
+    unquote [c] = (Unquoted, [c])
+    unquote s@(cl:tl) =
+      case reverse tl of
+        (cr:tr) | cr == cl && cl `elem` "\"'" -> (Quoted, reverse tr)
+        _ -> (if '\\' `elem` s then (Quoted, filter ((/=) '\\') s) else (Unquoted, s))
     -- Fun fact: bash considers << foo"" quoted, but not << <("foo").
-    -- Instead of replicating this, just read a token and strip quotes.
     readToken = do
         str <- readStringForParser readNormalWord
-        return (if any (`elem` quotes) str then Quoted else Unquoted,
-                filter (not . (`elem` quotes)) str)
-
+        return $ unquote str
 
 readPendingHereDocs = do
     docs <- popPendingHereDocs

